src/floyd_warshall.rs

```diff
@@ -0,0 +1,135 @@
+///
+/// # Examples
+/// ```rust
+/// use petgraph::{prelude::*, Graph, Directed};
+/// use petgraph::algo::floyd_warshall;
+/// use std::collections::HashMap;
+///
+/// let mut graph: Graph<(), (), Directed> = Graph::new();
+/// let a = graph.add_node(());
+/// let b = graph.add_node(());
+/// let c = graph.add_node(());
+/// let d = graph.add_node(());
+///
+/// graph.extend_with_edges(&[
+///    (a, b),
+///    (a, c),
+///    (a, d),
+///    (b, c),
+///    (b, d),
+///    (c, d)
+/// ]);
+///
+/// let weight_map: HashMap<(NodeIndex, NodeIndex), i32> = [
+///    ((a, a), 0), ((a, b), 1), ((a, c), 4), ((a, d), 10),
+///    ((b, b), 0), ((b, c), 2), ((b, d), 2),
+///    ((c, c), 0), ((c, d), 2)
+/// ].iter().cloned().collect();
+/// //     ----- b --------
+/// //    |      ^         | 2
+/// //    |    1 |    4    v
+/// //  2 |      a ------> c
+/// //    |   10 |         | 2
+/// //    |      v         v
+/// //     --->  d <-------
+///
+/// let inf = std::i32::MAX;
+/// let expected_res: HashMap<(NodeIndex, NodeIndex), i32> = [
+///    ((a, a), 0), ((a, b), 1), ((a, c), 3), ((a, d), 3),
+///    ((b, a), inf), ((b, b), 0), ((b, c), 2), ((b, d), 2),
+///    ((c, a), inf), ((c, b), inf), ((c, c), 0), ((c, d), 2),
+///    ((d, a), inf), ((d, b), inf), ((d, c), inf), ((d, d), 0),
+/// ].iter().cloned().collect();
+///
+///
+/// let res = floyd_warshall(&graph, |edge| {
+///     if let Some(weight) = weight_map.get(&(edge.source(), edge.target())) {
+///         *weight
+///     } else {
+///         inf
+///     }
+/// }).unwrap();
+///
+/// let nodes = [a, b, c, d];
+/// for node1 in &nodes {
+///     for node2 in &nodes {
+///         assert_eq!(res.get(&(*node1, *node2)).unwrap(), expected_res.get(&(*node1, *node2)).unwrap());
+///     }
+/// }
+/// ```
+pub fn floyd_warshall<G, F, K>(graph: G, mut edge_cost: F) -> Result<HashMap<(G::NodeId, G::NodeId), K>, NegativeCycle>
+    where
+        G: NodeCompactIndexable + IntoEdgeReferences + IntoNodeIdentifiers,
+        G::NodeId: Eq + Hash,
+        F: FnMut(G::EdgeRef) -> K,
+        K: BoundedMeasure + Copy,
+{
+    let num_of_nodes = graph.node_count();
+
+    // |V|x|V| matrix
+    let mut dist = vec![vec![K::max(); num_of_nodes]; num_of_nodes];
+
+    // init distances of paths with no intermediate nodes
+    for edge in graph.edge_references() {
+        dist[graph.to_index(edge.source())][graph.to_index(edge.target())] = edge_cost(edge);
+    }
+
+    // distance of each node to itself is 0(default value)
+    for node in graph.node_identifiers() {
+        dist[graph.to_index(node)][graph.to_index(node)] = K::default();
+    }
+
+    for k in 0..num_of_nodes {
+        for i in 0..num_of_nodes {
+            for j in 0..num_of_nodes {
+                // for an overflow: a + b > max: both values need to be positive and a > max - b must be satisfied
+                let overflow = dist[i][k] > K::default() && dist[k][j] > K::default() && dist[i][k] > K::max() - dist[k][j];
```

Comment:
You could use [overflowing_add](https://doc.rust-lang.org/std/primitive.usize.html#method.overflowing_add) to detect an over- or underflow. 
See [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=38c9d5ec22474795a7531631dbdaf628)